packages/react-router/modules/matchPath.js

```diff
@@ -25,37 +64,62 @@ function compilePath(path, options) {
 /**
  * Public API for matching a URL pathname to a path.
  */
-function matchPath(pathname, options = {}) {
-  if (typeof options === "string") options = { path: options };
+function matchPath(pathname, options = {}, parent) {
+  if (typeof options === "string") {
+    options = { path: options };
+  }
 
   const { path, exact = false, strict = false, sensitive = false } = options;
 
+  if (path == null) {
+    return parent;
+  }
+
   const paths = [].concat(path);
 
   return paths.reduce((matched, path) => {
-    if (matched) return matched;
+    if (matched) {
+      return matched;
+    }
+
+    const absolute = isAbsolute(path);
+
+    if (!absolute) {
+      path = resolvePath(path, parent && parent.url);
```

Comment:
If you use ```parent.pathname``` instead of ```parent.url```, you can simplify the param related code quite a bit, as path-to-regexp will most of the work for you.

Reply:
I am sure if I get your point here.

Reply:
You can shorten the rest of the function, if you use ```resolvePath(path, parent.path)```

```js
    if (!absolute) {
      path = resolvePath(path, parent && parent.path);
    }

    const { regexp, keys } = compilePath(path, {
      end: exact,
      strict,
      sensitive
    });
    const match = regexp.exec(pathname);

    if (!match) {
      return null;
    }

    const [url, ...values] = match;
    const isExact = pathname === url;

    if (exact && !isExact) {
      return null;
    }

    return {
      path, // the path pattern used to match
      url: path === "/" && url === "" ? "/" : url, // the matched portion of the URL
      isExact, // whether or not we matched exactly
      params: keys.reduce((memo, key, index) => {
        memo[key.name] = values[index];
        return memo;
      }, {})
   }

Reply:
That works, because match.path contains the original pattern used for matching, while URL contains the actual location.

```
path + parent.url === "/smith/:firstname"

path + parent.path === "/:lastname/:firstname"
```